graylog2-server/src/main/java/org/graylog/security/certutil/csr/ClientCertGenerator.java

```diff
@@ -0,0 +1,128 @@
+    private final CsrGenerator csrGenerator;
+    private final CsrSigner csrSigner;
+    private final ClusterConfigService clusterConfigService;
+    private final SecurityAdapter securityAdapter;
+    private final Path dataDir;
+
+    @Inject
+    public ClientCertGenerator(final CaService caService,
+                               @Named("password_secret") final String passwordSecret,
+                               @Named("data_dir") Path dataDir,
+                               final CsrGenerator csrGenerator,
+                               final CsrSigner csrSigner,
+                               final ClusterConfigService clusterConfigService,
+                               final SecurityAdapter securityAdapter) {
+        this.caService = caService;
+        this.passwordSecret = passwordSecret;
+        this.csrGenerator = csrGenerator;
+        this.csrSigner = csrSigner;
+        this.clusterConfigService = clusterConfigService;
+        this.securityAdapter = securityAdapter;
+        this.dataDir = dataDir;
+    }
+
+    private Path certFilePath(final String principal) {
+        return dataDir.resolve(Path.of(principal + ".cert"));
+    }
+
+    private String c(final Object o) throws IOException {
+        var writer = new StringWriter();
+        try (JcaPEMWriter jcaPEMWriter = new JcaPEMWriter(writer)) {
+            jcaPEMWriter.writeObject(o);
+        }
+        return writer.toString().replace("\n", "");
+    }
+
+    public ClientCert generateClientCert(final String principal,
+                                     final String role,
+                                     final char[] privateKeyPassword) throws ClientCertGenerationException {
+        try {
+            var renewalPolicy = this.clusterConfigService.get(RenewalPolicy.class);
+            var privateKeyEncryptedStorage = new PrivateKeyEncryptedFileStorage(certFilePath(principal));
+
+            final Optional<KeyStore> optKey = caService.loadKeyStore();
+            final var caKeystore = optKey.get();
+
+            var caPrivateKey = (PrivateKey) caKeystore.getKey(CA_KEY_ALIAS, passwordSecret.toCharArray());
+            var caCertificate = (X509Certificate) caKeystore.getCertificate(CA_KEY_ALIAS);
+
+            var csr = csrGenerator.generateCSR(privateKeyPassword, principal, List.of(principal), privateKeyEncryptedStorage);
+            var pk = privateKeyEncryptedStorage.readEncryptedKey(privateKeyPassword);
+            var cert = csrSigner.sign(caPrivateKey, caCertificate, csr, renewalPolicy);
+
+            securityAdapter.addUserToRoleMapping(role, principal);
+
+            return new ClientCert(principal, role, c(caCertificate), c(pk), c(cert));
+        } catch (Exception e) {
+            throw new ClientCertGenerationException("Failed to generate client certificate", e);
+        }
+    }
+
+    public void removeCertFor(final String role, final String principal) throws IOException {
+        var certFile = certFilePath(principal);
```

Comment:
If `principal` has multiple `role` values associated, those would collide, and we would remove the certificate for all of them.

I think it's safer to base the file name on the hash of `role + principal` and use the resulting string everywhere in this class, including when create the private key store.
I'm not sure how the UI is going to represent the client certificates that exist, so we might need to store the `role`, `principal` pair somewhere else, or get it from the role mapping. But I think that is an independent question from this PR.

Reply:
In general: how would using the file system work if you have multiple nodes, especially if those nodes might not be online at the same time?
I might be missing something but if this is a concern, the data needs to live somewhere else, probably in the database.

Reply:
This is the first draft - to make it right, I think we have to figure out where we're heading with the CA stuff:
- If it's a one time, throw away cert, we don't have to save the PK anywhere because we might not want to reuse it at all
- we could let the user create their own CSR, no need to generate a PK then
- roles etc. could be more than one, we should save that state to MongoDB - but again: I'd like to give it at least a little thought before I'm creating sth. that needs a migration in the future

Reply:
Ok, that makes sense. In that case, we should consider a follow-on PR that doesn't store the certificate at all (maybe post-beta, so we don't hold this up too much).
If we never read it again, it doesn't make much of a difference whether it lives on a single node.